### PR TITLE
Better instance outline from outside iframe

### PR DESCRIPTION
**Pull request by kof**
https://github.com/webstudio-is/webstudio-designer/issues/56

Instance outline is now rendered in the parent document and position/size is detected from canvas iframe.

This means it can't be broken by doing something unexpected on the canvas. It now also finds the best possible placement on top, inside or at the bottom of the instance, depending on the position and size of the instance.

### CODE REVIEW

**Comment by okonet** [comment]
So I'm testing it in the browser and I'm seeing weird things going on:

1. I think the selection should go away when clicking outside of the canvas. Now it stays on the last selected element.
2. The focus ring (?) from the Heading stays visible no matter what: CleanShot 2022-06-14 at 11 06 05@2x
3. The outline overlaps the element for me (Heading) 
image
4. The hover and selected elements doesn't play well together. 


I think that selection and hover (going to be selected) styles should be different. Figma uses different thickness of the outline, but we could experiment with different colors etc. I think.

**Review by okonet — COMMENTED** [review]
The code is LGTM but consider the `zIndex` comment and other mentioned issues.

**Comment by kof** [comment]
@okonet great review, working through the items now

**Comment by kof** [comment]
>1.  I think the selection should go away when clicking outside of the canvas. Now it stays on the last selected element.

I took that behavior from wf, but I am not sure what's the right one here. Usually, you don't have the typical unselect need, because almost everything you do requires you to select an element. I don't have a strong opinion on this, just not sure why we would want to unselect an element when clicking in workspace outside of canvas

**Comment by kof** [comment]
> 2. The focus ring (?) from the Heading stays visible no matter what:

Not sure I understand, the blue outline stays selected but you expected it to go away? Maybe a more detailed info or a loom would help understand better what you mean

**Comment by kof** [comment]
> 3. The outline overlaps the element for me (Heading)

This one is definitely a bug, I am not counting in the margin, fixing it now

**Comment by kof** [comment]
> 4. The hover and selected elements doesn't play well together.

Good catch. WF solves it by letting you click on the label and shows the labels "breadcrumb" basically. We could go the same route too, but maybe there are also other options? wdyt?

**Comment by kof** [comment]
> I think the use case I have in mind is to preview the work I've done without any overlays / outlines.

Great point actually, in wf you just click somewhere else on the canvas, but as you say it I agree, it should also work with click outside

**Comment by kof** [comment]
> 2. The focus ring (?) from the Heading stays visible no matter what:

Got this one reproduced now, all good, def. a bug

**Comment by okonet** [comment]
> Good catch. WF solves it by letting you click on the label and shows the labels "breadcrumb" basically. We could go the same route too, but maybe there are also other options? wdyt? 

I think that's not necessary for now (or maybe I don't understand what you mean specifically) but just making hovered element's outline / labels appear on top of currently selected with a slightly different style should solve this I think.

**Comment by taylornowotny** [comment]
Screen Shot 2022-06-14 at 7 31 48 AM

instance outlines cover up overlays

**Comment by kof** [comment]
The problem with hovering instance and having the label overlap with the label of the selected instance is an interesting problem, Webflow has it too ... I am tending to create a separate issue for that on after beta roadmap, wdyt @taylornowotny and @okonet ?


Screenshot 2022-06-14 at 23 39 28

**Comment by kof** [comment]
- [ ] hide outline in preview mode

**Comment by okonet** [comment]
I agree with a separate issue for this. I think that everything related to editing should be on a separate layer and not run in the iframe to have proper isolation. So my idea is to move this kind of designer-related UI to a separate layer similarly how this PR does for outlines. That would open the doors for things like direct manipulation and custom overlays. 

As an example I'm thinking about visualizing flex / grid like Google DevTools is later allowing direct manipulation of those:
CleanShot 2022-06-15 at 11 20 32@2x

WDYT?

**Comment by kof** [comment]
yes! I just forgot about text editor toolbar probably have to make another PR for it, otherwise this one is getting huge

File: app/designer/features/workspace/canvas-tools/outline/outline.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    left: 0,
+  },
+  {
+    variants: {
+      state: {
+        selected: {
+          zIndex: "$4",
+        },
+        hovered: {
+          zIndex: "$3",
```
[the remark this answers is not in the document]
Comment (kof) [reply app/designer/features/workspace/canvas-tools/outline/outline.tsx]: done

**Comment by kof** [comment]
Ok this PR is ready for a second round of review, known separate issues are:
- https://github.com/webstudio-is/webstudio-designer/issues/68
- https://github.com/webstudio-is/webstudio-designer/issues/69

**Comment by kof** [comment]
- [ ]  bug: when adding a new instance, it should be selected

**Review by okonet — APPROVED** [review]

**Pull request merged**